src/utilities/grading-helper.ts

```diff
@@ -0,0 +1,231 @@
+}
+
+export interface DetermineGradingRationaleResult {
+    isCompleted: boolean;
+    isExpired: boolean;
+    isLocked: boolean;
+    isWithinAttemptLimit: boolean;
+    isOnTime: boolean;
+    isLate: boolean;
+
+    willTrackAttemptReason: WillTrackAttemptReason;
+    willGetCreditReason: WillGetCreditReason;
+}
+
+export interface CalculateGradeOptions {
+    studentGrade: StudentGradeInterface;
+    topic: CourseTopicContentInterface;
+    question: CourseWWTopicQuestionInterface;
+
+    solutionDate: moment.Moment;
+
+    newScore: number;
+
+    timeOfSubmission: moment.Moment;
+}
+
+export interface CalculateGradeResult {
+    gradingRationale: DetermineGradingRationaleResult;
+    gradeUpdates: Partial<StudentGrade>;
+    score: number;
+}
+
+export const willBeGraded = (reason: WillGetCreditReason): boolean => {
+    return reason === WillGetCreditReason.YES || reason === WillGetCreditReason.YES_BUT_PARTIAL_CREDIT;
+};
+
+export const determineGradingRationale = ({
+    endDate,
+    deadDate,
+    locked,
+
+    overallBestScore,
+    numAttempts,
+
+    maxAttempts,
+
+    solutionDate,
+
+    timeOfSubmission
+}: DetermineGradingRationaleOptions): DetermineGradingRationaleResult => {
+    // use the same time for everything
+    // also if there was a case in which we were retroactively grading we could take an option param and swap this out
+    const theMoment = moment(timeOfSubmission);
+
+    const isCompleted = overallBestScore >= 1;
+    const isExpired = theMoment.isBefore(moment(solutionDate)) && theMoment.isSameOrAfter(moment(deadDate));
+    const areSolutionsAvailable = theMoment.isSameOrAfter(moment(solutionDate));
+
+    const isLocked = locked;
+    const isWithinAttemptLimit =
+        maxAttempts <= Constants.Course.INFINITE_ATTEMPT_NUMBER || // There is no limit to the number of attempts
+        numAttempts < maxAttempts; // They still have attempts left to use
+
+    const isOnTime = theMoment.isBefore(moment(endDate));
+    const isLate = theMoment.isBefore(moment(deadDate)) && theMoment.isSameOrAfter(moment(endDate));
+
+    let willTrackAttemptReason: WillTrackAttemptReason = WillTrackAttemptReason.UNKNOWN;
+    if (isCompleted) {
+        willTrackAttemptReason = WillTrackAttemptReason.NO_ALREADY_COMPLETED;
+    } else if (areSolutionsAvailable) {
+        willTrackAttemptReason = WillTrackAttemptReason.NO_IS_AFTER_SOLUTIONS_DATE;
+    }
+    // This should be the same as doing else, being extra explicity to be careful
+    // Also typescript complains if this is else because `UNKNOWN` is not possible below
+    else if (!isCompleted && !areSolutionsAvailable) {
+        willTrackAttemptReason = WillTrackAttemptReason.YES;
+    } else {
+        logger.error('An error occurred determining whether or not to keep the attempt');
+        willTrackAttemptReason = WillTrackAttemptReason.UNKNOWN;
+    }
+
+    let willGetCreditReason: WillGetCreditReason = WillGetCreditReason.UNKNOWN;
```

Comment:
I feel like this might be better as a conversion function.

Reply:
What do you mean by conversion function?
I feel like separating out the logic into its own function would be better too if that is what you are saying